sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/Changefeed.java

```diff
@@ -0,0 +1,138 @@
+    private final OffsetDateTime startTime; /* User provided start time. */
+    private final OffsetDateTime endTime; /* User provided end time. */
+    private OffsetDateTime lastConsumable; /* Last consumable time. The latest time the changefeed can safely be
+                                              read from.*/
+    private OffsetDateTime safeEndTime; /* Soonest time between lastConsumable and endTime. */
+    private final ChangefeedCursor cfCursor; /* Cursor associated with changefeed. */
+    private final ChangefeedCursor userCursor; /* User provided cursor. */
+    private final SegmentFactory segmentFactory; /* Segment factory. */
+
+    /**
+     * Creates a new Changefeed.
+     */
+    Changefeed(BlobContainerAsyncClient client, OffsetDateTime startTime, OffsetDateTime endTime,
+        ChangefeedCursor userCursor, SegmentFactory segmentFactory) {
+        this.client = client;
+        this.startTime = startTime;
+        this.endTime = endTime;
+        this.userCursor = userCursor;
+        this.segmentFactory = segmentFactory;
+
+        this.cfCursor = new ChangefeedCursor(this.endTime);
+        this.safeEndTime = endTime;
+    }
+
+    /**
+     * Get all the events for the Changefeed.
+     * @return A reactive stream of {@link BlobChangefeedEventWrapper}
+     */
+    Flux<BlobChangefeedEventWrapper> getEvents() {
+        return validateChangefeed()
+            .flatMap(ignore -> populateLastConsumable())
+            .flatMapMany(ignore -> listYears())
+            .concatMap(this::listSegmentsForYear)
+            .concatMap(this::getEventsForSegment);
+    }
+
+    /**
+     * Validates that changefeed has been enabled for the account.
+     */
+    private Mono<Boolean> validateChangefeed() {
+        return this.client.exists()
+            .flatMap(exists -> {
+                if (exists == null || !exists) {
+                    return Mono.error(new RuntimeException("Changefeed has not been enabled for this account."));
+                }
+                return Mono.just(true);
+            });
+    }
+
+    /**
+     * Populates the last consumable property from changefeed metadata.
+     * Log files in any segment that is dated after the date of the LastConsumable property in the
+     * $blobchangefeed/meta/segments.json file, should not be consumed by your application.
+     */
+    private Mono<OffsetDateTime> populateLastConsumable() {
+        /* We can keep the entire metadata file in memory since it is expected to only be a few hundred bytes. */
+        return DownloadUtils.downloadToString(this.client, METADATA_SEGMENT_PATH)
+            /* Parse JSON for last consumable. */
+            .flatMap(json -> {
+                try {
+                    ObjectMapper objectMapper = new ObjectMapper();
+                    JsonNode jsonNode = objectMapper.readTree(json);
+                    this.lastConsumable = OffsetDateTime.parse(jsonNode.get("lastConsumable").asText());
```

Comment:
Any chance this class could be used in multiple threads at once? Setting these class level variables could result in timing issues.

Reply:
Changfeed wasnt designed to be threadsafe just cause it's so dependent on all these states.

Reply:
Now that I'm looking at it again though - I'm gonna check to see if this is the only one that is stored like that and used later - cause I think the rest of the classes arent state dependent. I should be able to just pass lastConsumable down and make this part thread safe